### PR TITLE
refactor(container-definitions): Convert `AttachState` enum to const object

### PR DESCRIPTION
**Pull request by Josmithr**
We generally prefer const objects or simple unions over TS enums. This PR updates `AttachState` to use this preferred pattern. Note that while this is a breaking change type-wise (we are no longer exporting a TS enum), in practice it should be non-disruptive to consumers (note the update in the type-tests, which demonstrates compatibility of assignments in either direction between the new and old types).

### CODE REVIEW

File: packages/common/container-definitions/src/runtime.ts
Comment (alexvy86) [inline packages/common/container-definitions/src/runtime.ts:37]: Not pushing to do it here, it'd maybe be more disruptive; but I think I like `AttachStates` (in general, the plural) for the const object with the possible values, and `AttachState` (the singular) for the type that covers them all, to avoid complications during imports and writing TSDoc.

File: packages/common/container-definitions/src/loader.ts
Comment (alexvy86) [inline packages/common/container-definitions/src/loader.ts:162]:
```suggestion
	 * {@link (AttachState:variable).Attaching | attaching} to the Fluid service.
```

File: packages/common/container-definitions/src/loader.ts
Comment (alexvy86) [inline packages/common/container-definitions/src/loader.ts:331]:
```suggestion
	 * Will be undefined only when the container is in the {@link (AttachState:variable).Detached | detached} state.
```
The h-less Iron Lady would like a word 😝 .

File: packages/common/container-definitions/src/runtime.ts
Comment (Josmithr) [reply packages/common/container-definitions/src/runtime.ts:37]: Hmm. I can see some benefits to that pattern, though a downside is that it double the number of nominal exports, means that users have to reason about the relationship between the two (e.g. knowing that they can assign `AttachStates.Attached` to a parameter of type `AttachState`). Probably not unreasonable, and easily mitigated with docs.

**Review by alexvy86 — APPROVED** [review]
LGTM. Waiting for @anthony-murphy 's approval might be good, he's good at catching caveats with these kinds of changes.

File: packages/common/container-definitions/api-report/container-definitions.api.md
Comment (seanimam) [inline packages/common/container-definitions/api-report/container-definitions.api.md:448]: Does `typeof AttachState.Attaching` not resolve to "string"? or does it resolve to the string value in the const object?

Comment (alexvy86) [reply packages/common/container-definitions/api-report/container-definitions.api.md:448]: Yeah, I think the secret here is the `as const;` after the declaration, which makes the types of the object's properties not be "just some string", but "the specific string in the definition, and nothing else".

**Pull request closed without merging**
